File: Stage04-Play/Python/main.py
'''
Write the play() function
Write the take_action() function
Write the check_exit() function
'''
import time
import lib.kboard as kb #note use of lib. to locte file inside a subdirectory called lib
import lib.console as console
import narrator, game, shared, player, location

def check_exit(here:location.Location, direction:str) -> str:
	''' check if next location has an item_required to enter'''
	item_required:str = ""
	next_location:location.Location = here
	
	if direction == "to_north":
		next_location = shared.locations[here.to_north]
	elif direction == "to_east":
		next_location = shared.locations[here.to_east]
	elif direction == "to_south":
		next_location = shared.locations[here.to_south]
	elif direction == "to_west":
		next_location = shared.locations[here.to_west]
		
	item_required = next_location.item_required	
	if item_required != "":						# key or other item needed
		console.clear()
		if item_required in player.inventory: 	# is it in player inventory?
			print(f"You use the {item_required} in your inventory to enter...")
		else:
			print(f"You need {item_required} in your inventory to go that way")
			next_location = here
		time.sleep(2)
	
	return next_location.name	

def play() -> None:
	''' make choices about items in locations, inventory, move around etc '''
	action:str = ""
	row = console.clear()
	here:location.Location = shared.locations[shared.current_location]
	exits, row = here.display_location() # returns available exits and current console row
	action, param = take_action(here, exits, row)
	if action == "go":
		#check if next location has item_required
		shared.current_location = check_exit(here, param)		
	elif action == "examine":
		print(f"You examine the {param}:")
		print(f"{shared.items[param].get_description()}")
		time.sleep(2)
	elif action == "take":
		print(f"You take the {param} and put it in your backpack")
		player.add_to_inventory(param)
		here.remove_item(param)
		time.sleep(2)
	elif action == "inventory":
		# this action will only appear if player.inventory is not empty
		use_inventory()
	elif action == "quit":
		shared.gamestate = shared.gamestates['quit']

def	take_action(here:location.Location, exits:list, row:int) -> [str,str]:
	''' choose player action '''
	options:list = []
	# examine / take any items
	if len(here.items)> 0:
		for item in here.items:
			options.append(f"Examine {item}")
			options.append(f"Take {item}")
	# open inventory
	if len(player.inventory) > 0:
		options.append("Open your inventory")
	# take an exit
	if len(exits) > 0:
		for exit in exits:
			options.append(f"Go {exit}")

	# quit the game	
	options.append("Quit")
	choice:str = options[kb.menu("What next?", options, row)]

	if choice == "Quit":
		return "quit", ""
	elif "Go " in choice:
		# get direction 
		direction = f"to_{choice[3:]}" # to_north, to_east, etc
		return "go", direction
	elif "Examine" in choice:
		# get item
		item = choice[8:]
		return "examine", item
	elif "Take" in choice:
		item = choice[5:]
		return "take", item
	else:
		# open inventory
		return "inventory", ""

def use_inventory():
	choice = ""
	while choice != "Exit menu":
		row = console.clear()
		title = "Choose your option:"
		options = []
		for item in player.inventory:
			options.append(f"Examine {item}")
			options.append(f"Drop {item}")

		options.append("Exit menu")
		choice = options[kb.menu(title, options, row)]
		if "Examine" in choice:
			item = choice[8:]
			print(f"You examine the {item}")
			print(f"It is {shared.items[item].get_description()}")
			time.sleep(3)
		elif "Drop" in choice:
			item = choice[5:]
			print(f"You drop the {item}")
			shared.locations[shared.current_location].add_item(item)
			player.remove_from_inventory(item)
			time.sleep(3)

		if len(player.inventory) == 0:
			choice = "Exit menu"	

def main() -> None:
	''' Everything runs from here '''
	console.resize(80, 25, False)
	row = console.clear()	# If you do not use os.system('cls') before trying to set cursor position in kb methods it shows the escape codes!
	if shared.debug:														# allow disabling debug mode
		shared.debug = not kb.get_boolean("Debug mode is ON. Do you want to disable debug mode?", row)
		console.clear()
	
	if game.load_game():
		''' game loop '''
		shared.gamestate = shared.gamestates['play']							# start at 'play'
		while shared.gamestate == shared.gamestates['play']:					# continue game while in 'play' gamestate
			play()																# play the game. Gamestate can be changed in multiple places
		
		console.clear()	
		for message in narrator.end_message:
			print(message)
	else:
		print("Please correct errors in game locations before trying again")
  
	if console.is_console:
		input("Enter to quit")
	
main()

File: Stage04-Play/Lua/main.lua
--[[
Write the play() function
Write the takeAction() function
Write the checkExit() function
]]
kb 			= require"lib.kboard"
Console 	= require "lib.console"
Narrator 	= require "narrator"
Player 		= require "player"
Shared 		= require "shared"
Game 		= require "game"
--[[
global variables needed are dictionaries of Items, Enemies and Locations.
All global variables are stored in 'Shared.lua' and prefixed with Shared.
e.g. Shared.Items = {} holds a dictionary of Item objects available to any file via 'require "Shared"
]]

string.Split = Shared.Split
string.Trim = Shared.Trim
table.IsEmpty = Shared.IsEmpty

local function CheckExit(here, direction)
	--[[ check if next location has an itemRequired to enter ]]
	local nextLocation
	if direction == "north" then
		nextLocation = Shared.Locations[here:GetToNorth()]
	elseif direction == "east" then
		nextLocation = Shared.Locations[here:GetToEast()]
	elseif direction == "south" then
		nextLocation = Shared.Locations[here:GetToSouth()]
	elseif direction == "west" then
		nextLocation = Shared.Locations[here:GetToWest()]
	end	
	
	local itemRequired = nextLocation:GetItemRequired()	
	if itemRequired ~= "" then					-- key or other item needed
		if Shared.InList(Player.Inventory, itemRequired) then -- is it in player inventory?
			print("You use the "..itemRequired.." in your inventory to enter...")
		else
			print("You need "..itemRequired.." in your inventory to go that way")
			nextLocation = here
		end
		Shared.Sleep(2)
	end
	
	return nextLocation:GetName()	
end

local function TakeAction(here, exits, row)
	--[[ choose player action ]]
	local options = {}
	-- examine / take any items
	local items = here:GetItems()
	if #items > 0 then
		for _, item in ipairs(items) do
			table.insert(options, "Examine "..item)
			table.insert(options, "Take "..item)
		end
	end
	-- open inventory
	if #Player.Inventory > 0 then
		table.insert(options, "Open your inventory")
	end
	-- take an exit
	if #exits > 0 then
		for _, exit in pairs(exits) do
			table.insert(options, "Go "..exit)
		end
	end
	
	-- quit the game	
	table.insert(options, "Quit")
	local choice = options[kb.menu("What next?", options, row)]
	-- choice examples: "Attack...", Examine...", "Go..." etc
	if choice:find("Quit") ~= nil then
		return "quit", ""
	elseif choice:find("Go ") ~= nil then
		-- exit format "Go east"
		return "go", choice:sub(4)
	elseif choice:find("Examine") ~= nil then
		return "examine", choice:sub(9)
	elseif choice:find("Take") ~= nil then
		return "take", choice:sub(6)
	elseif choice:find("inventory") ~= nil then
		-- open inventory
		return "inventory", ""
	end
end

local function UseInventory()
	local choice = ""
	while choice ~= "Exit menu" do
		local row = Console.Clear()
		local title = "Choose your option:"
		local options = {}
		for _, item in pairs(Player.Inventory) do
			table.insert(options, "Examine "..item)
			table.insert(options, "Drop "..item)
		end
		table.insert(options, "Exit menu")
		choice = options[kb.menu(title, options, row)]
		if choice:find("Examine") ~= nil then
			local item = choice:sub(9)
			print("You examine the ".. item)
			print("It is "..Shared.Items[item]:GetDescription())
			if Shared.Items[item].type == "Weapon" then
				print("It is a Weapon with "..Shared.Items[item]:GetDamage().." hit points!")
			end
			Shared.Sleep(3)
		elseif choice:find("Drop") ~= nil then
			local item = choice:sub(6)
			print("You drop the ".. item)
			Shared.Locations[Shared.CurrentLocation]:AddItem(item)
			Player.RemoveFromInventory(item)
			Shared.Sleep(3)
		end
		if #Player.Inventory == 0 then
			choice = "Exit menu"
		end
	end
end

local function Play()
	--[[ make choices about items in locations, inventory, move around etc ]]
	local action, param
	local here = Shared.Locations[Shared.CurrentLocation]	-- Location object
	local exits = {}
	local row = Console.Clear()
	exits, row = here:DisplayLocation() 					-- returns available exits and current console row
	if not Console.IsConsole then row = -1 end				-- prevent 033[7;0H in ZeroBrane
	action, param = TakeAction(here, exits, row)
	--row = Console.Clear()
	if action == "go" then
		-- check if next location has item_required
		Shared.CurrentLocation = CheckExit(here, param)		
	elseif action == "examine" then
		print("You examine the "..param)
		print(Shared.Items[param]:GetDescription())
		Shared.Sleep(2)
	elseif action == "take" then
		print("You take the "..param.." and put it in your backpack")
		Player.AddToInventory(param)
		here:RemoveItem(param)
		Shared.Sleep(2)
	elseif action == "inventory" then
		-- this action will only appear if Player.Inventory is not empty
		UseInventory()
	elseif action == "quit" then
		Shared.Gamestate = Shared.Gamestates['quit']
	end
end

local function main()
	--[[ Everything runs from here ]]
	Console.Resize(80, 25, false)
	local row = Console.Clear()
	if Shared.Debug then							-- allow disabling debug mode
		Shared.Debug = not kb.getBoolean("Debug mode is ON. Do you want to disable debug mode?", row)
		row = Console.Clear()
	end
	if Game.LoadGame() then						-- load from file or use hard coded game if none exist
		--[[ game loop ]]
		Shared.Gamestate = Shared.Gamestates['play']							-- start at 'play'
		while Shared.Gamestate < Shared.Gamestates['quit'] do					-- continue game while in 'menu', 'play' gamestate
			Play()																-- play the game. Gamestate can be changed in multiple places
		end
		Console.Clear()	
		for _, message in ipairs(Narrator.EndMessage) do
			print(message)
		end
	else
		print("Please correct errors in game locations before trying again")
	end
	if Console.IsConsole then
		io.write("Enter to quit")
		io.read()
	end
end
-- Program runs from here:
main()
